libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp

```diff
@@ -0,0 +1,849 @@
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Code by Siddharth Bharat Purohit
+ */
+/*
+    Implements Common Flash Interface Driver based on following
+    standard published by JEDEC
+    * JEDEC Standard, JESD216, Serial Flash Discoverable Parameters (SFDP)
+*/
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_FlashIface_JEDEC.h"
+#include <AP_Math/AP_Math.h>
+
+extern const AP_HAL::HAL& hal;
+
+struct supported_device {
+    const char* name;
+    uint8_t manufacturer_id;
+    uint8_t device_id;
+};
+
+static const struct supported_device supported_devices[] = {
+    {"mt25q", 0x20, 0xBA}
+};
+
+// Vendor Specific Constants
+// Following Commands Sets were found here:
+// * JEDEC Standard JESD251-A1, Addendum No. 1 to JESD251, Optional x4 Quad I/O
+//   With Data Strobe
+/// NOTE: Except Read ID and Multiline Read ID, they seem to be
+// constant across manufacturers, but can't find official standard on
+// this.
+#define CMD_READ_ID                         0x9F
+#define CMD_MULTILINE_READ_ID               0xAF
+#define CMD_PAGE_PROGRAM                    0x02
+#define CMD_WRITE_DISABLE                   0x04
+#define CMD_READ_STATUS                     0x05
+#define CMD_MASS_ERASE                      0xC7
+#define CMD_RESET_ENABLE                    0x66
+#define CMD_RESET_MEMORY                    0x99
+#define CMD_READ_SFDP                       0x5A
+
+#define SFDP_MASK(lo, hi)                   (((1UL<<(hi)) - ((1UL<<(lo)))) + (1UL<<(hi)))
+#define SFDP_GET_BITS(x, lo, hi)            (((x) & SFDP_MASK(lo, hi)) >> (lo))
+#define SFDP_GET_BIT(x, bit)                ((x) & (1<<(bit)))
+
+#define SFDP_HDR_NUM_PARAMS(x)              (SFDP_GET_BITS(x[1], 16, 19) + 1)
+#define SFDP_HDR_PARAM_REV(x)               SFDP_GET_BITS(x[1], 0, 15)
+#define SFDP_PARAM_ID(x)                    ((SFDP_GET_BITS(x[0], 0, 3) << 8) | SFDP_GET_BITS(x[1], 24, 31))
+#define SFDP_PARAM_DWORD_LEN(x)             SFDP_GET_BITS(x[0], 24, 31)
+#define SFDP_PARAM_POINTER(x)               SFDP_GET_BITS(x[1], 0, 23)
+
+#define SFDP_REV_1_6                        0x0106
+
+#define Debug(fmt, args ...)  do {hal.console->printf("FlashIface_JEDEC: " fmt "\n", ## args);} while(0)
+
+#define MAX_READ_SIZE       1024UL
+
+bool AP_FlashIface_JEDEC::init()
+{
+    // Get device bus by name
+    _dev = nullptr;
+    for (uint8_t i = 0; i < ARRAY_SIZE(supported_devices); i++) {
+        _dev = hal.qspi->get_device(supported_devices[i].name);
+        if (_dev) {
+            _dev_list_idx = i;
+            break;
+        }
+    }
+    // Reset Device involves trying to soft reset the chip
+    // as when system reboots the device might not have.
+    reset_device();
+
+    // Detecting Device involves trying to read Device ID and matching
+    // with what we expect. Along with extracting info from SFDP
+    if (!detect_device()) {
+        hal.console->printf("Failed to detect flash device: %s\n", supported_devices[_dev_list_idx].name);
```

Comment:
Seems kind of serious, maybe a GCS error?

Reply:
No GCS is bootloader. 😆 even console is not present, so this should actually be Debug a print